### ISSUE TITLE
Incredible memory leak caused by rampant service checks

### ISSUE THREAD

**Issue by ThatJames** [hashicorp/consul#3457]
Consul is allowing the check to run indefinitely. This creates enormous leaks depending on the check's frequency (we run 10 seconds).

The checks in place definitely have their own issues, but that does not excuse the process being forgotten about and left to run amuck. Consul **must** be responsible for the entities it spawns. 

In 22 hours, just 2 simple scripts chewed 6.2 **gigs** of memory. 

```
systemctl output:
systemctl status consul
* consul.service - consul unit file
   Loaded: loaded (/etc/systemd/system/consul.service; disabled; vendor preset: enabled)
   Active: active (running) since Thu 2017-09-07 14:06:44 SAST; 22h ago
 Main PID: 24458 (consul)
    Tasks: 17752
   Memory: 6.2G
      CPU: 32min 58.694s
   CGroup: /system.slice/consul.service
````

### `consul version` for both Client and Server
Client: `n/a`
Server: `0.7.3`

###`consul info` for server (This was taken after reboot, apologies)
```
agent:
	check_monitors = 0
	check_ttls = 0
	checks = 0
	services = 1
build:
	prerelease = 
	revision = 'a90bb8f
	version = 0.7.3
consul:
	bootstrap = true
	known_datacenters = 1
	leader = true
	leader_addr = ***. ***. ***.66:8300
	server = true
raft:
	applied_index = 59321
	commit_index = 59321
	fsm_pending = 0
	last_contact = never
	last_log_index = 59321
	last_log_term = 7
	last_snapshot_index = 57345
	last_snapshot_term = 6
	latest_configuration = [{Suffrage:Voter ID:***. ***. ***.66:8300 Address: 
	latest_configuration_index = 1
	num_peers = 0
	protocol_version = 1
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Leader
	term = 7
runtime:
	arch = amd64
	cpu_count = 8
	goroutines = 57
	max_procs = 8
	os = linux
	version = go1.7.4
serf_lan:
	encrypted = false
	event_queue = 1
	event_time = 7
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 1
	members = 1
	query_queue = 0
	query_time = 1
serf_wan:
	encrypted = false
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 1
	members = 1
	query_queue = 0
	query_time = 1
```

### Operating system and Environment details
Ubuntu 16.04
Kernel: 4.4.0-92-generic

### Description of the Issue (and unexpected/desired result)
Consul allows checks to run indefinitely, creating a resource leak. I expect any checks still running on the next sweep to be marked as failed or ended. Not to be left running (and a new instance spawned).

### Reproduction steps
Create a check that does not end.
Define check in consul config directory.
Sit back and watch the resource space burn.

We only log error as info led to 200 gigs created in 3 days. No errors were reported.

**Comment by ThatJames** [comment]
Apologies if this was logged elsewhere, I did try looking for it by health checks and leaks but I didn't see anything describing my problem

**Comment by slackpad** [comment]
Hi @ThatJames there was an issue with checks running in Docker containers that was recently fixed but the check does run in a simple loop and should not accumulate subprocesses - https://github.com/hashicorp/consul/blob/v0.7.3/command/agent/check.go#L139-L154. Can you provide some more details about how your checks are configured?

**Comment by ThatJames** [comment]
Hi @slackpad 

Sorry for the late reply.

I had a look through the linked function, I am now confused as to how I am achieving this...

The check my side is definitely broken, but I suspect now I may have misunderstood your API...

Our check is pretty straight forward.

A shell file is called by consul that then executes a tiny in-house go binary to pull messages and determine the exit code. The app only reads the latest message but it's currently not partciluarly robust and does not time out. If a server is not there to answer it, it will just keep on keeping-on. 

What I've noticed now is that the instance count doesn't add up with regard to how long it runs. So this seems to happen somewhat intermittently. 

Consul config
```
{
  "service": {
    "name": "airtimeupdater",
    "tags": ["master"],
    "address": "127.0.0.1",
    "port": 8000,
    "enableTagOverride": false,
    "checks": [
      {
        "script": "/usr/local/bin/consulchecks/airtime_check.sh",
        "interval": "10s"
      }
    ]
  }
}
```
Script 
```
#!/bin/bash
/usr/local/bin/consulchecks/consulmon -servicename=airtimeupdater -config=/usr/local/bin/consulchecks/
```
Consulmon being our binary to tie consul into our monitoring platform. Really simple platform, asks a nats-io cluster for a message and then exists with that message's state (0, 1 or 2)

**Comment by ThatJames** [comment]
Bumping request.

I'm able to replicate this without a long running script. 

Can I ask for some direction with regard to a test case for you guys? I can't figure this out as I'm just a budding gopher. I agree that the simple loop should block on a process call, is there maybe something wrong with the way I've configured these checks? They look to be in line with the documentation

**Comment by slackpad** [comment]
Hi this got fixed in #3565 and released in Consul 1.0. Now we kill script checks that go over the timeout and wait for that to terminate, so we should never accumulate them.
